source/base/data_out_base.cc

```diff
@@ -8797,8 +8817,562 @@ DataOutBase::write_hdf5_parallel(
 
 #endif
 }
+#ifdef DEAL_II_WITH_NETCDF
+namespace
+{
+  /**
+   * Automatically close the netcdf file on scope exit.
+   */
+  struct ScopeNcClose
+  {
+    int ncid;
+    ~ScopeNcClose()
+    {
+      // Ignore the error code because nc_close can only fail
+      // if ncid is invalid, in which case nc_create/open or
+      // other functions already failed.
+      nc_close(ncid);
+    }
+  };
+
+  /**
+   * Helper function to create a NetCDF attribute for different data types.
+   */
+  int
+  nc_put_att_var(int                                         ncid,
+                 int                                         varid,
+                 const char                                 *name,
+                 const DataOutBase::CFFlags::AttributeValue &value)
+  {
+    return std::visit(
+      [&](const auto &v) {
+        using T = std::remove_cv_t<std::remove_reference_t<decltype(v)>>;
+        if constexpr (std::is_same_v<T, std::string>)
+          {
+            return nc_put_att_text(ncid, varid, name, v.size(), v.c_str());
+          }
+        else if constexpr (std::is_same_v<T, double>)
+          {
+            return nc_put_att_double(ncid, varid, name, NC_DOUBLE, 1, &v);
+          }
+        else if constexpr (std::is_same_v<T, int>)
+          {
+            return nc_put_att_int(ncid, varid, name, NC_INT, 1, &v);
+          }
+      },
+      value);
+  }
+
+  /**
+   * Helper function to write the mesh according to CF conventions.
+   */
+  template <int dim, int spacedim>
+  void
+  write_cf_mesh(const std::vector<DataOutBase::Patch<dim, spacedim>> &patches,
+                const DataOutBase::DataOutFilter &data_filter,
+                const DataOutBase::CFFlags       &flags,
+                const std::string                &filename,
+                const int                         global_node_cell_count[2],
+                const int                         global_node_cell_offsets[2],
```

Comment:
This isn't standard syntax (I'm surprised we don't get warnings on the CI): could you use `std::array` instead?